### PR TITLE
KAFKA-17131: Refactor TimeDefinitions

### PR DESCRIPTION
**Pull request by mjsax**
Refactor TimeDefintiions to not use old ProcessorContext any longer.

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/Suppressed.java [hunk 1]
```diff
@@ -72,16 +72,16 @@ static EagerBufferConfig maxBytes(final long byteLimit) {
         /**
          * Create a buffer unconstrained by size (either keys or bytes).
          *
-         * As a result, the buffer will consume as much memory as it needs, dictated by the time bound.
+         * <p>As a result, the buffer will consume as much memory as it needs, dictated by the time bound.
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Suppressed.java:75]: Just some side cleanup in related code...

**Review by cadonna — APPROVED** [review]
Thanks for the PR, @mjsax !

LGTM!

**Comment by mjsax** [comment]
I am a little "worried" about the failing test -- they are clearly unrelated, but should we fall back into a bad pattern and merge PRs w/o green build (on the other hand, it does not really buy us much to wait for these tests to get fixes...)

**Comment by mjsax** [comment]
Java 23:
```
Found 1 test failures:
FAILED ❌ KafkaAdminClientTest > testAdminClientApisAuthenticationFailure()
Found 3 flaky test failures:
FLAKY ⚠️  TransactionsWithTieredStoreTest > testReadCommittedConsumerShouldNotSeeUndecidedData(String, String).quorum=kraft.groupProtocol=consumer
FLAKY ⚠️  SslAdminIntegrationTest > "testCreateTopicsResponseMetadataAndConfig(String).quorum=kraft"
FLAKY ⚠️  DefaultStateUpdaterTest > shouldThrowIfAddingStandbyTasksWithSameId()
```
Java 17:
```
Found 1 test failures:
FAILED ❌ PlaintextAdminIntegrationTest > testConsumerGroupsDeprecatedConsumerGroupState(String, String).quorum=kraft.groupProtocol=consumer
Found 3 flaky test failures:
FLAKY ⚠️  TransactionsWithTieredStoreTest > "testBumpTransactionalEpochWithTV2Disabled(String, String, boolean).quorum=kraft, groupProtocol=classic, isTV2Enabled=false"
FLAKY ⚠️  TransactionsWithTieredStoreTest > "testBumpTransactionalEpochWithTV2Enabled(String, String, boolean).quorum=kraft, groupProtocol=consumer, isTV2Enabled=true"
FLAKY ⚠️  PlaintextAdminIntegrationTest > "testElectPreferredLeaders(String).quorum=kraft"
```

**Comment by cadonna** [comment]
> I am a little "worried" about the failing test -- they are clearly unrelated, but should we fall back into a bad pattern and merge PRs w/o green build

No, we should not fall back!

**Comment by mjsax** [comment]
Talked to David about the failing "Scan" task, and he told me, it's ok to merge w/o it passing. He did identify the problem already, and opened a PR to fix it: https://github.com/apache/kafka/pull/18272

**Comment by mjsax** [comment]
Merged to `trunk` and cherry-picked to `4.0`.

**Pull request merged**